### PR TITLE
feat(amplify_api): GraphQL Helpers Integration Tests

### PR DESCRIPTION
**Pull request by Equartey**
*Issue #, if available:*
Partial implementation of GraphQL model helper feature #308. 

*Description of changes:*
This PR introduces integration tests to validate the GraphQL helpers against a real endpoint. Includes tests for: `get`, `list`, `create`, `update`, and `delete`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_api/lib/src/graphql/model_mutations_factory.dart [hunk 1]
```diff
@@ -14,7 +14,7 @@ class ModelMutationsFactory extends ModelMutationsInterface {
 
   @override
   GraphQLRequest<T> create<T extends Model>(T model) {
-    Map<String, dynamic> variables = model.toJson();
+    Map<String, dynamic> variables = {"input": model.toJson()};
```
Comment (ragingsquirrel3) [inline packages/amplify_api/lib/src/graphql/model_mutations_factory.dart:17]: I have not looked at why unit tests are failing, but I would guess tests need to be updated to match this?

**Review by dnys1 — APPROVED** [review]
Looks good! Just fix up those failing tests and make the few changes I outlined and you'll be good to go 🚀

File: packages/amplify_api/example/integration_test/graph_ql_tests.dart [hunk 2]
```diff
@@ -44,12 +103,94 @@ void main() {
...
+      Blog blog = await addBlog(name);
+
+      var req = ModelMutations.delete(blog);
+      var _r = await Amplify.API.mutate(request: req);
+
+      var res = await _r.response;
+      Blog data = res.data;
+
+      expect(data.name, name);
+    });
```
Comment (ragingsquirrel3) [inline packages/amplify_api/example/integration_test/graph_ql_tests.dart:194]: suggestion: query the list of blogs before and after the delete to ensure the blog was actually deleted. Easiest way is to check the count of the result, ensure it went down by 1. Alternatively, you could query the list after the delete, iterate over all the blogs and ensure the blog not in the list result.

**Review by ragingsquirrel3 — CHANGES_REQUESTED** [review]

File: packages/amplify_api/example/integration_test/graph_ql_tests.dart [hunk 3]
```diff
@@ -12,24 +12,83 @@
...
+    }
+
     setUpAll(() async {
       if (!Amplify.isConfigured) {
-        await Amplify.addPlugins([AmplifyAPI()]);
+        await Amplify.addPlugins(
+            [AmplifyAPI(modelProvider: ModelProvider.instance)]);
         await Amplify.configure(amplifyconfig);
       }
+      if (blogIds.length == 0) {
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/amplify_api/example/integration_test/graph_ql_tests.dart]: Good catch, this is never touched and it's been omitted.

File: packages/amplify_api/example/integration_test/graph_ql_tests.dart [hunk 4]
```diff
@@ -12,24 +12,83 @@
...
+      var operation = await Amplify.API.mutate(request: req);
+
+      var response = await operation.response;
+      Map data = jsonDecode(response.data);
+
+      return true;
+    }
+
+    Future<bool> deleteAllBlogs() async {
+      blogIds.forEach((e) async {
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/amplify_api/example/integration_test/graph_ql_tests.dart]: Taking the delicate approach so we don't blow up someone's backend unexpectedly.

File: packages/amplify_api/example/integration_test/graph_ql_tests.dart [hunk 5]
```diff
@@ -44,12 +103,94 @@ void main() {
...
+      Blog blog_2 = await addBlog(blog_2_name);
+      Blog blog_3 = await addBlog(blog_3_name);
+
+      var req = ModelQueries.list<Blog>(Blog.classType);
+      var _r = await Amplify.API.query(request: req);
+
+      var res = await _r.response;
+      var data = res.data;
+
+      expect(data.items.length, greaterThanOrEqualTo(3));
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/amplify_api/example/integration_test/graph_ql_tests.dart]: I've updated the assertion.

File: packages/amplify_api/lib/src/graphql/graphql_response_decoder.dart [hunk 6]
```diff
@@ -48,6 +48,13 @@ class GraphQLResponseDecoder {
       dataJson = dataJson![element];
     });
 
+    if (dataJson == null) {
+      // TODO: T data is non-nullable, need to handle valid null response
+      throw ApiException('response from app sync was "null"',
+          recoverySuggestion:
+              "Current GraphQLResponse is non-nullable, please insure item exists before fetching");
```
Comment (ragingsquirrel3) [inline packages/amplify_api/lib/src/graphql/graphql_response_decoder.dart:55]: nit: insure -> ensure

**Review by ragingsquirrel3 — APPROVED** [review]
🚢  it w a typo. Nice job!

File: packages/amplify_api/lib/src/graphql/graphql_response_decoder.dart [hunk 7]
```diff
@@ -17,7 +17,7 @@ class GraphQLResponseDecoder {
       required List<GraphQLResponseError> errors}) {
     // if no ModelType fallback to default
     if (request.modelType == null) {
-      if (T == String) {
+      if (T == String || T == dynamic) {
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/amplify_api/lib/src/graphql/graphql_response_decoder.dart:20]: Will be addressing in another PR.

**Comment by rahul-readwhere** [comment]
Any comment on the release of this version to pub.dev ?

**Comment by ragingsquirrel3** [comment]
@rahul-readwhere the model helper branch needs to have some more updates and features before it is released to pub.dev. We are still working on subscriptions, getting/parsing nested models, query predicates and some helper methods related to paginated requests.

**Comment by rahul-readwhere** [comment]
@ragingsquirrel3 any time frame decided ? or will it depend on the features that will be added. Thanks for sharing :)

**Pull request merged**